netlib/http/semantics.py

```diff
@@ -12,8 +13,135 @@
 CONTENT_MISSING = 0
 
 
-class ProtocolMixin(object):
+class Headers(UserDict.DictMixin):
+    """
+    Header class which allows both convenient access to individual headers as well as
+    direct access to the underlying raw data. Provides a full dictionary interface.
+
+    Example:
+
+    .. code-block:: python
+
+        # Create header from a list of (header_name, header_value) tuples
+        >>> h = Headers([
+                ["Host","example.com"],
+                ["Accept","text/html"],
+                ["accept","application/xml"]
+            ])
+
+        # Headers mostly behave like a normal dict.
+        >>> h["Host"]
+        "example.com"
+
+        # HTTP Headers are case insensitive
+        >>> h["host"]
+        "example.com"
+
+        # Multiple headers are folded into a single header as per RFC7230
+        >>> h["Accept"]
+        "text/html, application/xml"
+
+        # Setting a header removes all existing headers with the same name.
+        >>> h["Accept"] = "application/text"
+        >>> h["Accept"]
+        "application/text"
+
+        # str(h) returns a HTTP1 header block.
+        >>> print(h)
+        Host: example.com
+        Accept: application/text
+
+        # For full control, the raw header lines can be accessed
+        >>> h.lines
+
+        # Headers can also be crated from keyword arguments
+        >>> h = Headers(host="example.com", content_type="application/xml")
+
+    Caveats:
+        For use with the "Set-Cookie" header, see :py:meth:`get_all`.
+    """
+
+    def __init__(self, lines=None, **headers):
+        """
+        For convenience, underscores in header names will be transformed to dashes.
+        This behaviour does not extend to other methods.
+
+        If ``**headers`` contains multiple keys that have equal ``.lower()``s,
+        the behavior is undefined.
+        """
+        self.lines = lines or []
```

Comment:
Good point. I set this public because we may need the very fine-grained control over lines at one or two places, and I found there's no harm exposing the raw stuff. Let's see - if we don't need it internall, we can hide it.